### PR TITLE
CLI interface added, function interface overhaul

### PR DESCRIPTION
**Pull request by nicmr**
## New
- New Command line interface - application can be used without the http interface. This will make writing cron jobs a breeze.

## Improved
- extracted shared functionality between cli and http handlers into separate functions
- improved CSV contents
- function interfaces, now pass time.Time structs instead of strings that still have to be parsed
- documentation, explanatory comments
- if-Chaining into switches
- better function names

## Documented
- How to use the filters
- Explanatory comments for difficult parts of code
- Documentation for cli and http interface

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -3,76 +3,146 @@ package main
...
 	w.Header().Set("Content-Disposition", "attachment; filename=costs.csv")
 	w.Write([]byte(output.CsvFileContent))
 
 }
 
 func main() {
-	http.HandleFunc("/costs", handleCosts)
-	log.Fatal(http.ListenAndServe(":8080", nil))
+	// Parse command line flags
+	var serve *bool = flag.Bool("serve", false, "If set, the program will respond to http requests at :8080 instead of just running once for a specific month")
```
Comment (markushinz) [inline main.go]: add port as flag

File: main.go [hunk 2]
```diff
@@ -3,76 +3,146 @@ package main
...
+			// Execute the request
+			output, err := costapi(parsedMonth)
+
+			if err != nil {
+				log.Println("GetCostAndUsageRequest failed", err)
+				os.Exit(1)
+			}
+
+			// Upload to S3
+			_, err = s3store.Upload(strings.NewReader(output.CsvFileContent), "bills/test_costs.csv")
```
Comment (markushinz) [inline main.go]: include current timestamp in filename to have unique filenames und to not override existing entries.

File: pkg/costs/aws.go [hunk 3]
```diff
@@ -40,43 +39,86 @@ func costexplorerCall(costexpl *(costexplorer.CostExplorer), start string, end s
...
 	output, err := costexpl.GetCostAndUsage(input)
 	if err != nil {
 		return nil, err
 	}
+	log.Println(output.String())
 	return output, nil
 }
 
-// CostsBetweenAWS call costexplorer after adding package-level variables as parameters,
+func maxGroupLen(arr []*costexplorer.ResultByTime) int {
```
Comment (markushinz) [inline pkg/costs/aws.go]: please add a description. What does this code?

File: pkg/costs/aws.go [hunk 4]
```diff
@@ -40,43 +39,86 @@ func costexplorerCall(costexpl *(costexplorer.CostExplorer), start string, end s
...
+
+	monthStr := month.Format(desiredFormat)
 
 	// Retrieve the required information for csvEntries from the output.
-	for index, element := range output.ResultsByTime {
-		csvEntries[index] = csv.Entry{
-			TimePeriodStart: *element.TimePeriod.Start,
-			TimePeriodEnd:   *element.TimePeriod.End,
-			Amount:          *element.Total["AmortizedCost"].Amount,
+	element := output.ResultsByTime[0]
```
Comment (markushinz) [inline pkg/costs/aws.go]: add a comment that this is the first month that we retrieve

Comment (nicmr) [reply pkg/costs/aws.go]: added!

File: pkg/costs/aws.go (hunk 3 above)
Comment (nicmr) [reply pkg/costs/aws.go]: Removed the code, as it's not needed after all :)

File: main.go (hunk 2 above)
Comment (nicmr) [reply main.go]: done!

File: README.md [hunk 5]
```diff
@@ -6,6 +6,21 @@ AWS billing service for Altemista Cloud
 
 Please read through [CONTRIBUTING.md](/CONTRIBUTING.md) before making any contributions.
 
+## Command line interface
+```
+~ altemista-billing --help
```
[the remark this answers is not in the document]
Comment (nicmr) [reply README.md]: indeed. fixed!

File: main.go (hunk 1 above)
Comment (nicmr) [reply main.go]: "--port" flag added

**Comment by nicmr** [comment]
Adressed all of markus' concerns and suggestions. Merging.

**Pull request merged**